src/plugins/discover/public/context_awareness/profiles_manager.ts

```diff
@@ -0,0 +1,170 @@
+  DataSourceProfile,
+  DocumentProfile,
+  RootProfileService,
+  DataSourceProfileService,
+  DocumentProfileService,
+  RootProfileProviderParams,
+  DataSourceProfileProviderParams,
+  DocumentProfileProviderParams,
+} from './profiles';
+
+interface SerializedRootProfileParams {
+  solutionNavId: RootProfileProviderParams['solutionNavId'];
+}
+
+interface SerializedDataSourceProfileParams {
+  dataViewId: string | undefined;
+  esqlQuery: string | undefined;
+}
+
+export interface GetProfilesOptions {
+  record?: DataTableRecord;
+}
+
+export class ProfilesManager {
+  private readonly rootProfile$ = new BehaviorSubject<ComposableProfile<RootProfile> | undefined>(
+    undefined
+  );
+  private readonly dataSourceProfile$ = new BehaviorSubject<
+    ComposableProfile<DataSourceProfile> | undefined
+  >(undefined);
+  private readonly documentProfiles$ = new BehaviorSubject<
+    Map<string, ComposableProfile<DocumentProfile>>
+  >(new Map());
+
+  private prevRootProfileParams?: SerializedRootProfileParams;
+  private prevDataSourceProfileParams?: SerializedDataSourceProfileParams;
+  private rootProfileAbortController?: AbortController;
+  private dataSourceProfileAbortController?: AbortController;
+
+  constructor(
+    private readonly rootProfileService: RootProfileService,
+    private readonly dataSourceProfileService: DataSourceProfileService,
+    private readonly documentProfileService: DocumentProfileService
+  ) {}
+
+  public async resolveRootProfile(params: RootProfileProviderParams) {
+    const serializedParams = this.serializeRootProfileParams(params);
+
+    if (this.rootProfile$.getValue() && isEqual(this.prevRootProfileParams, serializedParams)) {
+      return;
+    }
+
+    const abortController = new AbortController();
+    this.rootProfileAbortController?.abort();
+    this.rootProfileAbortController = abortController;
+
+    const profile = await this.rootProfileService.resolve(params);
+
+    if (abortController.signal.aborted) {
+      return;
+    }
+
+    this.rootProfile$.next(profile);
+    this.prevRootProfileParams = serializedParams;
+  }
+
+  public async resolveDataSourceProfile(params: DataSourceProfileProviderParams) {
+    const serializedParams = this.serializeDataSourceProfileParams(params);
+
+    if (
+      this.dataSourceProfile$.getValue() &&
+      isEqual(this.prevDataSourceProfileParams, serializedParams)
+    ) {
+      return;
+    }
+
+    const abortController = new AbortController();
+    this.dataSourceProfileAbortController?.abort();
+    this.dataSourceProfileAbortController = abortController;
+
+    const profile = await this.dataSourceProfileService.resolve(params);
+
+    if (abortController.signal.aborted) {
+      return;
+    }
+
+    this.dataSourceProfile$.next(profile);
+    this.prevDataSourceProfileParams = serializedParams;
+  }
+
+  public createDocumentProfilesCollector() {
+    const documentProfiles = new Map<string, ComposableProfile<DocumentProfile>>();
```

Comment:
I have to admit I'm a bit concerned about building this parallel mapping. It introduces even more hidden state. This state needs to be kept in sync with other state in Discover, which sounds risky. I wonder if we can't just store the resolved context on the record and make the lookup from context to profile cheap for the consumers using a profile id. What do you think?

Reply:
Yeah that's fair. My thinking here was to store all the active profiles in a central place, but I agree the separate state introduces some risk of getting out of sync, and some extra code complexity to manage.

Is your suggestion to provide a profile ID when registering a provider and inject it into the resolved context, then add a get profile by ID to the profile service? Something along these lines:
```ts
// Consumer
documentProfileService.register({
  profileId: 'logProfile',
  profile: { ... },
  resolve: () => { ... }
});

// ProfilesManager.resolveDocumentProfile({ record })
const { context, profileId } = documentProfileService.resolve({ record });
const contextWithId = { ...context, profileId  };

record.context = contextWithId;

// ProfilesManager.getProfiles({ record })
const profile = documentProfileService.getProfileById(record.context.profileId);
```

Reply:
Exactly, if the profile gets an id upon registration the `context` could contain that for cheap lookup as shown in your code sample. Do you see a problem with making the profile id a common property on all context interfaces?

Reply:
I switched to this approach instead of collecting the document profiles in a separate map: https://github.com/elastic/kibana/pull/183797/commits/d8351ca9b511be5338740760c1ac3d22a5b3c3e5. It'll also be easier now to access the current contexts from `ProfilesManager`, which will be helpful for things like debugging in the inspector. Let me know if you think this seems like a better approach.